### ISSUE TITLE
How can I get directories and files in hierarchical structure from bucket ?

### ISSUE THREAD

**Issue by oscarpires** [aws/aws-sdk-net#1758]
## The Question
I know about the **ListObjectsV2Request** used to list all objects in a bucket. But he will return "/Dir/Subdir/file.txt" for each object, I'd like to make a hierarchical struture for directories and theirs subfolders or files.
I don't identifiy in C# API how to get only dirs and then files from that directory.
Please what's the best way to recursively create a files hierarchical struture ??
thanks a lot for any help

### Environment
* SDK Version: 3.5.6.2
* Package Version: AWSSDK.S3
* OS Info: Win10
* Build Environment VS2017
* Targeted .NET Platform:  framework 4.9

---

This is a :question: directory hierarchy of a bucket

**Comment by ashishdhingra** [comment]
Hi @oscarpires,

Good morning.

As per article https://docs.aws.amazon.com/AmazonS3/latest/user-guide/using-folders.html, 
```
In Amazon S3, buckets and objects are the primary resources, and objects are stored in buckets. Amazon S3 has a flat structure instead of a hierarchy like you would see in a file system. However, for the sake of organizational simplicity, the Amazon S3 console supports the folder concept as a means of grouping objects. Amazon S3 does this by using a shared name prefix for objects (that is, objects have names that begin with a common string). Object names are also referred to as key names.
```

I understand you want to write logic to create a hierarchical structure from S3 bucket similar to Windows. Following references could be helpful:

- AWS .NET SDK BCL Custom code for `S3DirectoryInfo` and `S3FileInfo` available at https://github.com/aws/aws-sdk-net/tree/master/sdk/src/Services/S3/Custom/_bcl/IO. 

- Stackoverflow post for listing folders https://stackoverflow.com/questions/33545065/amazon-s3-how-to-get-a-list-of-folders-in-the-bucket

Basically you would need to write logic and use appropriate data structure to walk through the objects to construct hierarchy based on a prefix.

Hope this helps. Let me know if this issue could be closed.

Thanks,
Ashish
